---
title: "Check strands with liftOver"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages

```{r}
suppressPackageStartupMessages({
    library(here)
    library(Rsamtools)
    library(rtracklayer)
    library(VariantAnnotation)
    library(BSgenome.Hsapiens.UCSC.hg38)
    library(BSgenome.Ptroglodytes.UCSC.panTro5)
    library(BSgenome.GGorilla.UCSC.gorGor5)
    library(BSgenome.PAbelii.UCSC.ponAbe2)
})
```

## Subset the vcf file on chromosome 8

```{r}
gr <- GRanges("chr8", IRanges(1, 145138636))
gr <- unlist(slidingWindows(gr, width = 100, step = 100))
set.seed(24)
gr <- sort(sample(gr, 10000))

vcf <- readVcf(
    TabixFile(here("raw_data/ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz")),
    param = ScanVcfParam(which = gr)
)

rg <- rowRanges(vcf)
rg$QUAL <- NULL
rg$paramRangeID <- NULL

strand(rg) <- "+"
rg
```

## Set the Reference


## Function to do liftOver

```{r}
lift <- function(gr, chain) {
    if (is.character(chain))
        chain <- import.chain(chain)
    
    grl.target <- local({
        mcols(gr) <- NULL
        liftOver(gr, chain)
    })
    grl.target
}
queryString <- function(grl, bsgenome) {
    gr <- unlist(unname(grl))
    dna <- DNAStringSet(Views(bsgenome, gr))
    relist(dna, unname(grl))
}
```


```{r}
rg$loc.panTro5 <- lift(rg, here("raw_data/chainfiles/hg38ToPanTro5.over.chain"))
rg$dna.panTro5 <- queryString(rg$loc.panTro5, BSgenome.Ptroglodytes.UCSC.panTro5)
#rg$loc.gorGor5 <- lift(rg, here("raw_data/chainfiles/hg38ToGorGor5.over.chain"))
#rg$dna.gorGor5 <- queryString(rg$loc.panTro5, BSgenome.GGorilla.UCSC.gorGor5)
rg$loc.ponAbe2 <- lift(rg, here("raw_data/chainfiles/hg38ToPonAbe2.over.chain"))
rg$dna.ponAbe2 <- queryString(rg$loc.ponAbe2, BSgenome.PAbelii.UCSC.ponAbe2)
```

# Plot

```{r}
# TMP: trim the range

rg_flat <- rg[(lengths(rg$dna.panTro5) == 1) & (lengths(rg$dna.ponAbe2) == 1)]
lengths(rg_flat$dna.panTro5) %>% table
rg_flat$dna.panTro5 <- unlist(rg_flat$dna.panTro5)
rg_flat$dna.ponAbe2 <- unlist(rg_flat$dna.ponAbe2)
rg_flat

data <- tibble::data_frame(
    pos = start(rg_flat),
    match.panTro5 =  rg_flat$REF == rg_flat$dna.panTro5,
    match.ponAde2 =  rg_flat$REF == rg_flat$dna.ponAbe2
    #consit.ancestor =  rg_flat$dna.panTro5 == rg_flat$dna.ponAbe2
)
data

sample_data <- data[sample(seq(nrow(data)), 10000),]

library(tidyr)
library(dplyr)
library(ggplot2)
sample_data %>%
    tidyr::gather(key = "MatchWho", value = "value", -pos) %>%
    ggplot() +
    geom_point(aes(x = pos, y = value, color = MatchWho), size = 0.1,
               position = position_jitter(height = 0.3, width = 0),
               alpha = 0.4)
```






