---
title: "Interpreting One-to-many Liftover Results"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


The implementation of `rtracklayer::liftOver` and UCSC liftover tool is different.
See https://support.bioconductor.org/p/45159/.

The following is quoted from the UCSC Genome Browser mailing list (https://groups.google.com/a/soe.ucsc.edu/forum/#!topic/genome/ffskDdFf6gA) as a comparision
between liftOver and pslMap.

> liftOver maps the start and end positions independently, so if there is a rearrangement between the two sequences, the number of bases in the region can change dramatically. pslMap forces a block structure on the alignment so the number of bases in the alignment is the same in both sequences (though there may be very large "introns", and inversions and translocations will cause blocks to be dropped because they aren't in the same order and orientation any more).
>
> I prefer pslMap because it's a base to base mapping. liftOver is useful if you just want to know where the same region is in a
>
> different assembly of the same individual which presumably isn't rearranged with itself but may have inserted bases where there was a gap before (though obviously translocations do occur within cells of individuals...hello cancer). Both procedures use the same chain or psl file, which can be made by lastz or blat or etc.The author of pslMap also had this to say:
>
> Both methods are dependent on the alignments used and how they are filtered. The README in pslMap describes how to do the syntenic filtering used for cross-species transmap. It requires some experimentation to see if the alignments produce the desired results.
>
> The bit algorithmic difference is that liftover lifts block boundaries which pslMap is per-base. So liftover produces something more resembling an annotation while pslMap shows the evolutionary changes. Also, pslMap requires converting to psl; there are tools to do this.

The rtracklayer's `liftOver` is more like pslMap which considers possible rearrangements
between the species. A range can be mapped to multiple ranges -- sum of width of these
ranges should equal to width of the original range.


## Investigation on chromosome 8

```{r}
suppressPackageStartupMessages({
    library(here)
    
    library(Rsamtools)
    library(rtracklayer)
    library(VariantAnnotation)

    library(BSgenome.Hsapiens.UCSC.hg38)
    library(BSgenome.PAbelii.UCSC.ponAbe2)
})
```


Sample some ranges from chr8

```{r}
range <- as(seqinfo(BSgenome.Hsapiens.UCSC.hg38), "GRanges")
range <- range["chr8"]
seqlevels(range) <- "chr8"
range <- slidingWindows(range, width = 1000, step = 1000)
range <- range$chr8
range <- unname(range)

set.seed(5)
sample_ranges <- sample(range, 1000)
sample_ranges
```

```{r}
sample_variants <- readVcf(
    TabixFile(here("raw_data/ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz")),
    param = ScanVcfParam(which = sample_ranges)
)
sample_variants <- rowRanges(sample_variants)
sample_variants$paramRangeID <- NULL
sample_variants$QUAL <- NULL
sample_variants$FILTER <- NULL
strand(sample_variants) <- "+"
sample_variants
```

### LiftOver the sample variants to ponAbe2


```{r}
grl.ponAbe2 <- liftOver(
    sample_variants,
    import.chain(here("raw_data/chainfiles/hg38ToPonAbe2.over.chain"))
)
grl.ponAbe2@unlistData$REF <- NULL
grl.ponAbe2@unlistData$ALT <- NULL
sample_variants$loc.ponAbe2 <- grl.ponAbe2
```

## Fact Checks

### Number of the mapped ranges

```{r}
table(lengths(sample_variants$loc.ponAbe2))
```

### One-base range does not map to multiple ranges

```{r}
all(lengths(sample_variants[width(sample_variants) == 1]$loc.ponAbe2) %in% c(1, 0))
```

### Width of the original range is always greater than sum of width of the mapped ranges

```{r}
mulmapping_variants <- sample_variants[lengths(sample_variants$loc.ponAbe2) > 1]
table(width(mulmapping_variants) - sapply(mulmapping_variants$loc.ponAbe2, function(x) sum(width(x))))
```

### Chromosomes and strand of the mapped ranges

```{r}
chr_strand <- function(x) paste(seqnames(x), as.character(strand(x)), sep = ":")
table(chr_strand(unlist(sample_variants$loc.ponAbe2)))
table(chr_strand(unlist(mulmapping_variants$loc.ponAbe2)))
```

### Mapped ranges are on the same strand

```{r}
is_same_strand <- function(x) length(unique(as.character(strand(x)))) <= 1
all(sapply(mulmapping_variants$loc.ponAbe2, is_same_strand))
```

### There is no overlap between the mapped ranges

```{r}
no_overlap <- function(x) {
    all(countOverlaps(x, x) == 1)
}
all(sapply(mulmapping_variants$loc.ponAbe2, no_overlap))
```

### The mapped ranges are mostly ordered from 5' to 3'

```{r}
in_order <- function(x) {
    strand <- unique(as.character(strand(x)))
    stopifnot(length(strand) == 1)
    stopifnot(length(x) >= 2)
    
    for (i in seq_along(x)) {
        if (i == 1) next
        a <- x[i - 1]
        b <- x[i]
        if (strand == "+")
            if (start(b) <= end(a))
                return(FALSE)
        if (strand == "-")
            if (start(b) >= end(a))
                return(FALSE)
    }
    return(TRUE)
}
table(sapply(mulmapping_variants$loc.ponAbe2, in_order))
Filter(Negate(in_order), mulmapping_variants$loc.ponAbe2)
```



## TODO

generate a table showing the mapped sequences


